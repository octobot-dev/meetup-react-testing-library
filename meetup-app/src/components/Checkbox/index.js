import React from 'react';
import PropTypes from 'prop-types';

import CheckboxContainer from './styles/CheckboxContainer';
import HiddenCheckbox from './styles/HiddenCheckbox';
import StyledCheckbox from './styles/StyledCheckbox';
import Icon from './styles/Icon';

function Checkbox({ checked, onChange, label }) {
  return (
    <CheckboxContainer>
      <HiddenCheckbox checked={checked} onChange={onChange} aria-label={label} />
      <StyledCheckbox checked={checked} onClick={onChange}>
        <Icon viewBox="0 0 24 24">
          <polyline points="20 6 9 17 4 12" />
        </Icon>
      </StyledCheckbox>
    </CheckboxContainer>
  );
}

Checkbox.propTypes = {
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
};

export default Checkbox;
