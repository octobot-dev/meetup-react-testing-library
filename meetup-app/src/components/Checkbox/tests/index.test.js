import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import 'jest-styled-components';

import Checkbox from '../index';

describe('Checkbox', () => {
  it('renders correctly when unchecked', () => {
    const mockOnChange = jest.fn();
    const { container } = render(<Checkbox checked={false} onChange={mockOnChange} label="My checkbox" />);
    // Snpshot test for component
    expect(container).toMatchSnapshot();

    const checkbox = screen.getByRole('checkbox', { name: 'My checkbox' }); // Will fail if checkbox role is not present
    expect(checkbox).not.toBeChecked();
  });

  it('renders correctly when checked', () => {
    const mockOnChange = jest.fn();
    const { container } = render(<Checkbox checked onChange={mockOnChange} label="My checkbox" />);
    // Snpshot test for component
    expect(container).toMatchSnapshot();

    const checkbox = screen.getByRole('checkbox', { name: 'My checkbox' }); // Will fail if checkbox role is not present
    expect(checkbox).toBeChecked();
  });

  it('calls onChange callback when changed', () => {
    const mockOnChange = jest.fn();
    render(<Checkbox checked={false} onChange={mockOnChange} label="My checkbox" />);

    const checkbox = screen.getByRole('checkbox', { name: 'My checkbox' });
    expect(checkbox).not.toBeChecked();
    userEvent.click(checkbox);
    expect(mockOnChange).toHaveBeenCalled();
  });
});
