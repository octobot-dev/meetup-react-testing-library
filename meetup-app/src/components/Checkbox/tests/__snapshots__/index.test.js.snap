// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly when checked 1`] = `
.c0 {
  display: inline-block;
  vertical-align: middle;
  cursor: pointer;
  width: 24px;
  height: 24px;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  opacity: 0;
}

.c4 {
  fill: none;
  stroke: white;
  stroke-width: 2px;
}

.c2 {
  display: inline-block;
  width: 24px;
  height: 24px;
  background: salmon;
  border-radius: 4px;
  -webkit-transition: all 150ms;
  transition: all 150ms;
}

.c2:hover {
  background: salmon;
}

.c2 .c3 {
  visibility: visible;
}

<div>
  <div
    class="c0"
  >
    <input
      aria-label="My checkbox"
      checked=""
      class="c1"
      type="checkbox"
    />
    <div
      class="c2"
    >
      <svg
        class="c3 c4"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Checkbox renders correctly when unchecked 1`] = `
.c0 {
  display: inline-block;
  vertical-align: middle;
  cursor: pointer;
  width: 24px;
  height: 24px;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  opacity: 0;
}

.c4 {
  fill: none;
  stroke: white;
  stroke-width: 2px;
}

.c2 {
  display: inline-block;
  width: 24px;
  height: 24px;
  background: #fcb797;
  border-radius: 4px;
  -webkit-transition: all 150ms;
  transition: all 150ms;
}

.c2:hover {
  background: salmon;
}

.c2 .c3 {
  visibility: hidden;
}

<div>
  <div
    class="c0"
  >
    <input
      aria-label="My checkbox"
      class="c1"
      type="checkbox"
    />
    <div
      class="c2"
    >
      <svg
        class="c3 c4"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
  </div>
</div>
`;
